flax/cursor.py

```diff
@@ -0,0 +1,341 @@
+
+  def __setitem__(self, key, value):
+    self.changes[key] = Cursor(value, ParentKey(self, key))
+
+  def __setattr__(self, name, value):
+    self.changes[name] = Cursor(value, ParentKey(self, name))
+
+  def set(self, value) -> A:
+    """Set a new value for an attribute, property, element or entry
+    in the Cursor object and return a copy of it.
+
+    Example::
+
+      from flax.cursor import cursor
+      from flax.training import train_state
+      import optax
+
+      dict_obj = {'a': 1, 'b': (2, 3), 'c': [4, 5]}
+      modified_dict_obj = cursor(dict_obj)['b'][0].set(10)
+      assert modified_dict_obj == {'a': 1, 'b': (10, 3), 'c': [4, 5]}
+
+      state = train_state.TrainState.create(
+          apply_fn=lambda x: x,
+          params=dict_obj,
+          tx=optax.adam(1e-3),
+      )
+      modified_state = cursor(state).params['b'][1].set(10)
+      assert modified_state.params == {'a': 1, 'b': (2, 10), 'c': [4, 5]}
+
+    Args:
+      value: the value used to set an attribute, property, element or entry in the Cursor object
+    Returns:
+      A copy of the Cursor object with the new set value.
+    """
+    if self.parent_key is None:
+      return value
+    parent, key = self.parent_key.parent, self.parent_key.key  # type: ignore
+    parent.changes[key] = value
+    return parent.root.build()
+
+  def build(self) -> A:
+    """Create and return a copy of the Cursor object with accumulated changes.
+    This method is to be called after making changes to the Cursor object.
+
+    NOTE: The new object is built bottom-up, the changes will be first applied
+    to the leaf nodes, and then its parent, all the way up to the root.
+
+    Example::
+
+      from flax.cursor import cursor
+      from flax.training import train_state
+      import optax
+
+      dict_obj = {'a': 1, 'b': (2, 3), 'c': [4, 5]}
+      c = cursor(dict_obj)
+      c['b'][0] = 10
+      c['a'] = (100, 200)
+      modified_dict_obj = c.build()
+      assert modified_dict_obj == {'a': (100, 200), 'b': (10, 3), 'c': [4, 5]}
+
+      state = train_state.TrainState.create(
+          apply_fn=lambda x: x,
+          params=dict_obj,
+          tx=optax.adam(1e-3),
+      )
+      new_fn = lambda x: x + 1
+      c = cursor(state)
+      c.params['b'][1] = 10
+      c.apply_fn = new_fn
+      modified_state = c.build()
+      assert modified_state.params == {'a': 1, 'b': (2, 10), 'c': [4, 5]}
+      assert modified_state.apply_fn == new_fn
+
+    Returns:
+      A copy of the Cursor object with the accumulated changes.
+    """
+    changes = {
+        key: child.build() if isinstance(child, Cursor) else child
+        for key, child in self.changes.items()
+    }
+    if isinstance(self.obj, FrozenDict):
+      obj = self.obj.copy(changes)  # type: ignore
+    elif isinstance(self.obj, (dict, list)):
+      obj = self.obj.copy()  # type: ignore
+      for key, value in changes.items():
+        obj[key] = value
+    elif isinstance(self.obj, tuple):
```

Comment:
we should also support setting attributes by name on a NamedTuple - these can only be ducktyped by looking for a `_fields` attribute.  A lot of libraries still use namedtuples over dataclasses.